### PR TITLE
Entropy: nRF: Clear NVIC flags before CPU idling

### LINKED ISSUE #26105: Test kernel.memory_protection.stack_random fails on nrf52dk_nrf52832
**Describe the bug**
Sanitycheck fails when running kernel.memory_protection.stack_random test on nrf52dk_nrf52832
(.zephyr/tests/kernel/mem_protect/stack_random). The test ends in timeout because no proper serial output is detected.

**To Reproduce**
1. run in zephyr dir ```./scripts/sanitycheck --testcase-root ./tests/kernel/mem_protect/stack_random --device-testing -p nrf52dk_nrf52832 -v -v --device-serial /dev/ttyACM0```
2. See error

**Expected behavior**
The test passes on nrf52dk_nrf52832

**Impact**
not clear

**Screenshots or console output**
```Flashing nrf52dk_nrf52832
-- west flash: using runner nrfjprog
-- runners.nrfjprog: Flashing file: /home/maciej/zephyrproject2/zephyr/sanity-out/nrf52dk_nrf52832/tests/kernel/mem_protect/stack_random/kernel.memory_protection.stack_random/zephyr/zephyr.hex
Parsing hex file.
Erasing page at address 0x0.
Erasing page at address 0x1000.
Erasing page at address 0x2000.
Erasing page at address 0x3000.
Erasing page at address 0x4000.
Erasing page at address 0x5000.
Erasing page at address 0x6000.
Applying system reset.
Checking that the area to write is not protected.
Programming device.
Enabling pin reset.
Applying pin reset.
-- runners.nrfjprog: Board with serial number 682923426 flashed successfully.
make[3]: Leaving directory '/home/maciej/zephyrproject2/zephyr/sanity-out/nrf52dk_nrf52832/tests/kernel/mem_protect/stack_random/kernel.memory_protection.stack_random'
[100%] Built target flash
make[2]: Leaving directory '/home/maciej/zephyrproject2/zephyr/sanity-out/nrf52dk_nrf52832/tests/kernel/mem_protect/stack_random/kernel.memory_protection.stack_random'
make[1]: Leaving directory '/home/maciej/zephyrproject2/zephyr/sanity-out/nrf52dk_nrf52832/tests/kernel/mem_protect/stack_random/kernel.memory_protection.stack_random'
make: Leaving directory '/home/maciej/zephyrproject2/zephyr/sanity-out/nrf52dk_nrf52832/tests/kernel/mem_protect/stack_random/kernel.memory_protection.stack_random'

DEBUG   - Timed out while monitoring serial output on nrf52dk_nrf52832
INFO    - 1/1 nrf52dk_nrf52832          tests/kernel/mem_protect/stack_random/kernel.memory_protection.stack_random FAILED Timeout (device 62.971s)
ERROR   - see: /home/maciej/zephyrproject2/zephyr/sanity-out/nrf52dk_nrf52832/tests/kernel/mem_protect/stack_random/kernel.memory_protection.stack_random/build.log

INFO    - 0 of 1 tests passed (0.00%), 1 failed, 0 skipped with 0 warnings in 67.84 seconds
INFO    - In total 1 test cases were executed on 1 out of total 253 platforms (0.40%)
INFO    - 1 tests executed on platforms, 0 tests were only built.

Hardware distribution summary:

| Board            | ID   |   Counter |
|------------------|------|-----------|
| nrf52dk_nrf52832 |      |         1 |
```

**Environment:**
 - OS: Linux
 - Toolchain Zephyr SDK
 - Commit SHA 1b1d728a18

**Additional context**
The test passes on other Nordic platforms (passes on nrf9160dk_nrf9160 and nrf52840dk_nrf52840).

### PR DESCRIPTION
**Pull request by ioannisg**
Fixes #26105 

Aims at fixing an (ugly) race condition in the entropy driver.

Tested on
- nrf51dk_nrf51422
- nrf52dk_nrf52832
- nrf52833dk_nrf52833
- nrf52840dk_nrf52840

### CODE REVIEW

File: drivers/entropy/entropy_nrf5.c [hunk 1]
```diff
@@ -278,6 +278,12 @@ static int entropy_nrf5_get_entropy_isr(struct device *dev, uint8_t *buf, uint16
 		nrf_rng_event_clear(NRF_RNG, NRF_RNG_EVENT_VALRDY);
 		nrf_rng_task_trigger(NRF_RNG, NRF_RNG_TASK_START);
 
+		/* Clear NVIC pending bit. This ensures that a subsequent RNG event
+		 * will set the Cortex-M single-bit event register to 1 (the bit is
+		 * set when NVIC pending IRQ status is changed from 0 to 1)
+		 */
+		NVIC_ClearPendingIRQ(IRQN);
```
Comment (pabigot) [inline drivers/entropy/entropy_nrf5.c:285]: Should this be before the trigger?

Comment (ioannisg) [reply drivers/entropy/entropy_nrf5.c:285]: Well, that's probably the "intuitive thing to do", although it, clearly, has no impact. That's because: 
- the task could be already triggered but a previous invocation - we make no assumptions the task is triggered here for the first time
- as long as it _becomes_ clear _after_ we have triggered the task and _before_ we enter the while loop containing WFE, we are certain that the event will occur

No strong opinion though

**Comment by ioannisg** [comment]
@aescolar @wopu-ot could you assist here? I provided a very simple model for ARM's DSB intrinsic as "no-op" for nrf52_bsim, I wonder if this should be sufficient. Effectively a DSB makes sure memory writes are complete before executing next instruction, so it should not matter for the simulated platform (but would like you to take a look)

**Review by cvinayak — APPROVED** [review]
LGTM

File: boards/posix/nrf52_bsim/irq_handler.c [hunk 2]
```diff
@@ -390,3 +390,8 @@ void __SEV(void)
 {
 	CPU_event_set_flag = true;
 }
+
+/* Model ARM Data Synchronization Barrier instruction as no-op */
+void __DSB(void)
```
Comment (wopu-ot) [inline boards/posix/nrf52_bsim/irq_handler.c:395]: The no-op should be sufficient for nrf52_bsim. __DMB is defined in the HW model (see https://github.com/BabbleSim/ext_NRF52_hw_models/blob/master/src/nrfx/hal/nrf_soc_if.h#L98) . For consistency, I would prefer __DSB to come from there too.

Comment (aescolar) [reply boards/posix/nrf52_bsim/irq_handler.c:395]: I agree with @wopu-ot that this one does not belong here (the `__WFE` `__WFI` are here because of the model of the event behaviour)
But I'm thinking that maybe the `__DMB` and `__NOP` should not have been placed in the HW models, but in a new `boards/posix/nrf52_bsim/cmsis.h` as macros, and have this `cmsis.h` be included from `board_soc.h`. (This __<asm_inst> macros/functions are defined by CMSIS)

@wopu-ot what do you think about also moving the DMB and NOP to a new cmsis.h?

Comment (ioannisg) [reply boards/posix/nrf52_bsim/irq_handler.c:395]: @aescolar @wopu-ot thanks for your input on this. I would probably prefer to see the functions in some cmsis.h as macros or static inline empty functions, what do you think?

Comment (wopu-ot) [reply boards/posix/nrf52_bsim/irq_handler.c:395]: @aescolar @ioannisg I would be fine with moving DMB and NOP, my point was more that DSB should be next to DMB.

Comment (aescolar) [reply boards/posix/nrf52_bsim/irq_handler.c:395]: @ioannisg Would you then create that new file with these definitions, and send a PR to the HW models removing those 2 static inlines from there?
@rugeGerritsen I guess you are ok with __DMB and __NOP being in that new cmsis.h?

Comment (rugeGerritsen) [reply boards/posix/nrf52_bsim/irq_handler.c:395]: @aescolar . I'm fine with moving those to cmsis.h

Comment (ioannisg) [reply boards/posix/nrf52_bsim/irq_handler.c:395]: @aescolar do get this PR pass CI, I suggest I do this for DSB and we can move the DMP, NOP, in the cmisis (and remove from BabbleSIm) in a subsequent PR, what do you think?

Comment (aescolar) [reply boards/posix/nrf52_bsim/irq_handler.c:395]: @ioannisg it can be. Would it be you the kind person preparing that 2nd PR? :)

Comment (ioannisg) [reply boards/posix/nrf52_bsim/irq_handler.c:395]: sure, can do that

**Review by anangl — APPROVED** [review]

**Review by wopu-ot — APPROVED** [review]

File: boards/posix/nrf52_bsim/irq_handler.c (hunk 2 above)
Comment (ioannisg) [reply boards/posix/nrf52_bsim/irq_handler.c]: @aescolar @wopu-ot https://github.com/BabbleSim/ext_NRF52_hw_models/pull/6
I guess, if you merge that one, then we can send a PR in Zephyr to add these functions in cmsis.h

**Review by aescolar — APPROVED** [review]

File: drivers/entropy/entropy_nrf5.c (hunk 1 above)
Comment (pabigot) [reply drivers/entropy/entropy_nrf5.c:286]: OK

**Review by pabigot — COMMENTED** [review]
No objections, but it's far enough out of of my comfort zone that I don't feel qualified to give an explicit approval.

File: drivers/entropy/entropy_nrf5.c [hunk 3]
```diff
@@ -278,11 +278,27 @@ static int entropy_nrf5_get_entropy_isr(struct device *dev, uint8_t *buf, uint16
 		nrf_rng_event_clear(NRF_RNG, NRF_RNG_EVENT_VALRDY);
 		nrf_rng_task_trigger(NRF_RNG, NRF_RNG_TASK_START);
 
+		/* Clear NVIC pending bit. This ensures that a subsequent
+		 * RNG event will set the Cortex-M single-bit event register
+		 * to 1 (the bit is set when NVIC pending IRQ status is
+		 * changed from 0 to 1)
+		 */
+		NVIC_ClearPendingIRQ(IRQN);
```
Comment (carlescufi) [inline drivers/entropy/entropy_nrf5.c:286]: So what happens if the ISR executes between these 2? It doesn't matter because we have the `nrf_rng_event_check()` later?

Comment (ioannisg) [reply drivers/entropy/entropy_nrf5.c:286]: @carlescufi the problem we try to solve here is the following: 
When the `get_entropy_isr()` function is called with the RNG interrupt masked, that is
- either IRQs are (globally)  locked, or
- the RNG IRQ is not enabled, or
- the function is called from a higher handler priority

then it is possible that the NVIC pending flag of the RNG interrupt is already set (to 1) before we do the first WFE. This means that the ARM core won't set the single-bit event register when the next RNG number is ready, since this is done when the NVIC pending IRQ flag switches from 0 to 1. This event register might be set due to other reasons, of course, but **if it is not, then the MCU cannot exit WFE when the next RNG number is ready**

> So what happens if the ISR executes between these 2?

Actually, the IRQ is disabled on line 274, so it cannot execute between the clear and the trigger. But it can be set to pending, which will set the event register and allow us to exit WFE. 

To summarize:
By adding an NVIC_ClearPendingIRQ(IRQN); statement before the first WFE, I guarantee that we will always witness a pending flag switch from 0 to 1, and this event will set the event-register bit to 1, allowing to exit WFE.

Comment (ioannisg) [reply drivers/entropy/entropy_nrf5.c:286]: @stephanosio maybe you could also take a quick look and review this reasoning?

**Review by carlescufi — APPROVED** [review]
